### PR TITLE
Enable to restore functions, sequences and views with schema level re…

### PR DESCRIPTION
**Pull request by royc1**
…store filter

for gprestore -S option.

Add behave tests and unit tests.

Authors: Chumki Roy & Marbin Tan

### CODE REVIEW

File: gpMgmt/bin/gppylib/test/behave/mgmt_utils/backup.feature [hunk 1]
```diff
@@ -3553,6 +3553,18 @@ Feature: Validate command line arguments
         Then gpdbrestore should return a return code of 0
         And verify that there are "2190" tuples in "bkdb" for table "public.foo4"
 
+    Scenario: Scheme level restore with gprestore -S option for views, sequences, and functions
```
Comment (kaknikhil) [inline gpMgmt/bin/gppylib/test/behave/mgmt_utils/backup.feature]: /s/Scheme/Schema and /s/gprestore/gpdbrestore

File: gpMgmt/bin/gprestore_filter.py [hunk 2]
```diff
@@ -161,25 +168,28 @@ def process_schema(dump_schemas, dump_tables, fdin, fdout, change_schema=None, s
             output = False
             function_ddl = False
             passedDropSchemaSection = True
-            if type in ['TABLE', 'EXTERNAL TABLE']:
+
+            if type in ['SCHEMA']:
+                # Make sure that schemas are created before restoring the desired tables.
+                output = check_valid_schema(name, dump_schemas, schema_level_restore_list)
+            elif type in ['TABLE', 'EXTERNAL TABLE', 'VIEW', 'SEQUENCE']:
```
Comment (Chibin) [inline gpMgmt/bin/gprestore_filter.py]: Probably need to remove these prints here

File: gpMgmt/bin/gppylib/test/behave/mgmt_utils/backup.feature [hunk 3]
```diff
@@ -3553,6 +3553,18 @@ Feature: Validate command line arguments
...
+        Given the test is initialized
+        And the user runs "psql -f gppylib/test/behave/mgmt_utils/steps/data/schema_level_test_workload.sql template1"
+        When the user runs command "gpcrondump -a -x schema_level_test_db"
+        Then gpcrondump should return a return code of 0
+        And the timestamp from gpcrondump is stored
+        When the user runs "gpdbrestore -S s1 -a -e" with the stored timestamp
```
Comment (kaknikhil) [inline gpMgmt/bin/gppylib/test/behave/mgmt_utils/backup.feature]: do we need a step for testing gpdbrestore without -S to make sure that it restores all types of objects without -S?

File: gpMgmt/bin/gprestore_filter.py [hunk 4]
```diff
@@ -161,25 +168,28 @@ def process_schema(dump_schemas, dump_tables, fdin, fdout, change_schema=None, s
...
+                print "TYPE"
```
Comment (kaknikhil) [inline gpMgmt/bin/gprestore_filter.py]: debug print statements??

File: gpMgmt/bin/gppylib/test/behave/mgmt_utils/backup.feature (hunk 3 above)
Comment (khuddlefish) [inline gpMgmt/bin/gppylib/test/behave/mgmt_utils/backup.feature]: This test does not belong to the story. But if we want to test that with the full backup and restore, we can create a story to add those tests with full backup and restore.

-Chumki & Karen

File: gpMgmt/bin/gprestore_filter.py [hunk 5]
```diff
@@ -161,25 +168,28 @@ def process_schema(dump_schemas, dump_tables, fdin, fdout, change_schema=None, s
...
                 further_investigation_required = False
-                output = check_valid_table(schema, name, dump_tables, schema_level_restore_list)
-                if output:
-                    search_path = True
+                output = check_valid_relname(schema, name, dump_tables, schema_level_restore_list)
             elif type in ['CONSTRAINT']:
                 further_investigation_required = True
-                if (dump_schemas and schema in dump_schemas) or (schema_level_restore_list and schema in schema_level_restore_list):
-                    line_buff = line 
+                if check_valid_schema(schema, dump_schemas, schema_level_restore_list):
```
Comment (kaknikhil) [inline gpMgmt/bin/gprestore_filter.py]: should the variable "output" be assigned in constraint as well ?

Comment (royc1) [inline gpMgmt/bin/gprestore_filter.py]: if further_investigation_required = True then later in the code it does set search_path = True. Therefore we don't need to add the variable "output" to set search_path = True

**Comment by royc1** [comment]
Thanks for review.

**Pull request closed without merging**